docs: Campaigns updates (#8510)
sourcegraph/sourcegraph: doc/user/automation.md

# Campaigns

> > Campaigns is currently in private beta for select enterprise customers.

Reviewer: ```suggestion
> Campaigns are currently in private beta for select enterprise customers.
```

See [discussion here](https://sourcegraph.slack.com/archives/CMMTWQQ49/p1582186187246300)

1. Go to `/campaigns` on your Sourcegraph instance and click on the "New campaign" button
2. Fill in a name for the campaign and a description
3. Create the campaign
4. Track changesets by adding them to the campaign through the form on the Campaign page

## Creating a campaign using the src CLI

If you have not already, first [install](https://github.com/sourcegraph/src-cli), [set up and configure](https://github.com/sourcegraph/src-cli#setup) the `src` CLI to point to your Sourcegraph instance.

To create a campaign via the CLI:

1. Create an action JSON file (e.g. `action.json`) that contains an action definition
1. _Optional_: See repositories the action would run over: `src actions scope-query -f action.json`
1. Create a set of patches by executing the action over repositories: `src actions exec -f action.json > patches.json`
1. Save the patches in Sourcegraph by creating a campaign plan based on these patches: `src campaign plan create-from-patches < patches.json`
1. Create a campaign from the campaign plan: `src campaigns create -branch=<branch-name> -plan=<plan-ID-returned-by-previous-command>`

Read on detailed steps and documentation.

## Where to best run campaigns

> The patches for a campaign plan are generated on the machine where the `src` CLI is executed, which in turn, downloads zip archives and runs each step against each repository. Therefore, we recommend campaign plans should run (in most cases) on a Linux machine with considerable CPU, RAM, and network bandwidth to reduce the execution time. Putting this machine in the same network as your Sourcegraph instance will also improve performance.

Reviewer: At this point in the document the user doesn't know what a "campaign plan" is yet, so I would rephrase:

```suggestion
The patches for a campaign are generated on the machine where the `src` CLI is executed, which in turn, downloads zip archives and runs each step against each repository. For most usecases we recommend that `src` CLI should be run on a Linux machine with considerable CPU, RAM, and network bandwidth to reduce the execution time. Putting this machine in the same network as your Sourcegraph instance will also improve performance.
```

> Another factor affecting execution time is the number of jobs executed in parallel, which is by default, the number of cores on the machine. This can be adjusted using the `-j` parameter.

Reviewer: ```suggestion
Another factor affecting execution time is the number of jobs executed in parallel, which is by default the number of cores on the machine. This can be adjusted using the `-j` parameter.
```

To make it simpler for customers, we're [working on remote execution of campaign](https://github.com/sourcegraph/src-cli/pull/128) of campaign actions and would love your feedback.

> ## Defining a campaign action plan

Reviewer: ```suggestion
## Defining a action
```

Author: ```suggestion
## Defining an action
```

> The first thing we need is a definition of an "action". An action contains a list of steps to run commands or Docker containers to produce a patch for each repository returned by the results of the `scopeQuery` search string.

Reviewer: ```suggestion
The first thing we need is a definition of an "action". An action contains a list of steps to run in each repository returned by the results of the `scopeQuery` search string.
```

The next sentence already explains the types of steps.

}
```

> This action will execute on every repository that has `go-` in its name and doesn't have an `INSTALL.md` file.

Reviewer: ```suggestion
Due to the definition of the `"scopeQuery"` this action will execute in every repository that has `go-` in its name and doesn't have an `INSTALL.md` file.
```

1. The first step (a `command` step) creates an `INSTALL.md` file in the root directory of each repository by running `sh` in a temporary copy of each repository. **This is executed on the machine on which `src` is being run.** Note that the first element in `"args"` is the command itself.

2. The second step, again a `"command"` step, runs the `sed` command to replace text in the `README.md` file in the root of each repository (the `-i ''` argument is only necessary for BSD versions of `sed` that usually come with macOS). Please note that the executed command is simply `sed` which means its arguments are _not_ expanded, as they would be in a shell. To achieve that, execute the `sed` as part of a shell invocation (using `sh -c` and passing in a single argument, for example, like in the first step).

3. The third step builds a Docker image from the specified `"dockerfile"` and starts a container with this image in which the repository is mounted under `/work`.

> 4. The fourth step pulls the `golang:1.13-alpine` image from Docker hub, starts a container from it and runs `go fix /work/...` in it.

Reviewer: Just realized that we don't do this yet:

```suggestion
4. The fourth step starts a Docker container based on the `golang:1.13-alpine` image and runs `go fix /work/...` in it.
```

As you can see from these examples, the "output" of an action is the modified, local copy of a repository.

Save that definition in a file called `action.json` (or any other name of your choosing).

## Executing an action to produce patches

With our action file defined, we can now execute it:

```sh
src actions exec -f action.json
```

This command is going to:

> 1. Pull the required Docker image if necessary.

Reviewer: Again: we don't pull images automatically yet

```suggestion
1. Build the required Docker image if necessary.
```

## Updating a campaign

You can also apply a new campaign plan for an existing campaign. Following the creation of the campaign plan with the `src campaign plan create-from-patches` command, a URL will be output that will guide you to the web UI to allow you to change an existing campaign's campaign plan.

On this page, click "Preview" for the campaign that will be updated. From there, the delta of existing and new changesets will be displayed. Click "Update" to finalize the proposed changes.

Edits to the name and description of a campaign can also be made in the web UI with the changes reflected in each changeset. The branch name of a draft campaign with a plan can also be edited, but only if the campaign doesn't contain any published changesets.

## Clearing the campaign action cache

Campaign diffs are intelligently cached based on the `scopeQuery` and defined `steps`, but the need to clear the cache to run the steps from scratch may be required.

> Clearing the cache requires the manual deletion of the cache directory. If no `-cache` flag is passed to `src actions exec`, the default location of the cache is used which can be found for your platform by running:

Reviewer: ```suggestion
Clearing the cache requires either manually emptying the cache directory or using a different one. If no `-cache` flag is passed to `src actions exec`, the default location of the cache is used which can be found for your platform by running:
```

Author: I think this conflates clearing (deleting) the cache and forcing a fresh run by specifying a new and empty cache location but that's a minor concern and isn't worth rewriting at this point. Thanks for the addition.

```sh

> src actions exec -help 2>&1 | head -n 3

Reviewer: ```suggestion
src actions exec -help
```

The other thing is brittle, since it could break with new flags.

```

## Example campaigns

> The following examples demonstrate various campaign types for different languages using both commands and Docker images. They also provide commentary on considerations such as adjusting the duration (`-timeout`) for actions that exceed the 15 minute default limit.

Reviewer: ```suggestion
The following examples demonstrate various types of campaigns for different languages using both commands and Docker images. They also provide commentary on considerations such as adjusting the duration (`-timeout`) for actions that exceed the 15 minute default limit.
```
